include/kernel.h

```diff
@@ -0,0 +1,534 @@
+ * API.
+ *
+ * A context-switch between threads can happen in the following situations:
+ *
+ *   * Preemptively: when a hardware interrupt generates a event that
+ *     wakes up a thread (see #kevent). This is the most common scenario:
+ *     an event connected to an interrupt is fired, and a higher priority
+ *     thread waiting for the event is woken up.
+ * 
+ *   * Cooperatively: when a thread explicitly calls a blocking function
+ *     like #kthread_sleep or #kthread_wait_event to wait for an event.
+ *     The kernel will put the current thread to sleep and schedule the
+ *     highest-priority thread among the ready ones, including threads
+ *     that have a priority lower than the one that goes to sleep. When
+ *     the event is fired, the lower-priority thread will be preempted
+ *     to give back control to the thread that just woke up.
+ *
+ *   * Cooperatively: when a thread explicitly calls #kthread_yield to
+ *     leave the CPU to other threads that might be ready at the same time.
+ *     Notice that, given the hard real-time guarantee, this only makes
+ *     sense if there are multiple ready threads at the same priority level,
+ *     as calling #kthread_yield will not schedule a lower-priority thread.
+ *     It is normally not necessary to call #thread_yield.
+ *
+ * In addition to threads (#kthread_t) and events (#kevent_t), the kernel
+ * library offers a message passing communication primitive called "message
+ * box" (#kmbox_t). A mailbox is a circular buffer of messages called
+ * "mails"; any number of threads can try to send mails into a mailbox,
+ * and any number of threads can try to read mails from a mailbox. Once a
+ * thread receives a mail from a mailbox, the mail is removed and other
+ * threads will not see it anymore. 
+ *
+ * Mailboxes can be useful as communicative primitives among threads
+ * (eg: to send tasks to other threads), or a synchronization primitive
+ * to wait for multiple events (see #kmbox_attach_event).
+ *
+ * @{
+ */
+
+/**
+ * @brief a kernel thread for parallel execution
+ *
+ * This structure represents a thread that can be scheduled
+ * for parallel execution. Create a thread with #kthread_new,
+ * allocating the required memory for the stack.
+ *
+ * Normally, there's no need to explicitly manage the lifetime
+ * or state of a thread. Once the thread is created, it is
+ * immediately started, and the kernel will scheduled it when
+ * required. If the thread exits (by simply returning by its
+ * entry point function), it will be cleaned up and the memory
+ * released.
+ *
+ * #kthread_new returns a pointer to the thread (a pointer
+ * to kthread_t) that can be used to externally manage the
+ * thread like changing its priority or killing it.
+ */
+typedef struct kthread_s
+{
+	/** Pointer to the top of the stack, which contains the thread register state */
+	reg_block_t* stack_state;
+	/** Current state of interrupt depth for this thread */
+	int interrupt_depth;
```

Comment:
This variable might be made smaller as well.

Reply:
Yeah but that is a snapshot of the global state (to fake TLS) and the global state is int. Changing it here would require changing the global state. I can easily think that 255 nesting levels ought to be enough, but I'd rather not mess with it in the context of this PR.

Reply:
So should we understand that scheduling a thread which has its interrupts enabled will enable the interrups? I haven't checked the implementation yet on purpose and this made me also think we may need to keep track of all global state per thread e.g `interrupt_disabled_tick`?

Reply:
> So should we understand that scheduling a thread which has its interrupts enabled will enable the interrupts?

Well, in general, this is not a user-visible behavior. Most code runs with interrupts enabled. If the user disable interrupts, it's usually to run some short critical code that cannot be interrupted, and in fact with interrupts disabled the kernel scheduler can not run! So disabling interrupts also has the effect of disabled thread preemption, which is a good default of existing code (in modern kernels, disabling interrupts and disabling preemptions are two different concepts, but in our simple case I think it's OK to conflate the two).

In the kernel implementation itself, though we need to force thread switching while interrupts are disabled. This is just an internal implementation detail, but when we do, we absolutely need to restore enabled interrupts for scheduled threads (and then restore disabled interrupts when we get back to the switching point!).

> we may need to keep track of all global state per thread e.g `interrupt_disabled_tick`

In the specific case of `interrupt_disabled_tick`, that is OK to keep as a global state as `timer_ticks` uses it just as a "new enough" point in time. But in general yes, we might of course find out in future that there is some global state that should actually be thread-local storage. I would say this is unlikely when dealing with hardware libraries that deal with hardware peripherals (as the peripheral is just one and has just one state, irrespective of the running thread), but it's not impossible. If we find out other thread-local storage, we would have to deal with it. We can't stuff everything in kthread_t, but for now it's a fast, simple way.